File: pms-backend/src/main/java/com/pms/service/ProjectPhaseService.java
package com.pms.service;

import com.pms.dto.ProjectPhaseRequest;
import com.pms.dto.ProjectPhaseResponse;
import com.pms.entity.*;
import com.pms.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class ProjectPhaseService {
    private final ProjectPhaseRepository phaseRepository;
    private final PhaseMilestoneRepository milestoneRepository;
    private final MilestoneActivityRepository activityRepository;
    private final ProjectDefinitionRepository projectRepository;

    @Transactional
    public ProjectPhaseResponse createPhase(Long projectId, ProjectPhaseRequest request) {
        // Validate project exists
        ProjectDefinition project = projectRepository.findById(projectId)
                .orElseThrow(() -> new RuntimeException("Project not found"));

        // Validate phase weight
        if (request.getPhaseWeight() < 0 || request.getPhaseWeight() > 100) {
            throw new RuntimeException("Phase weight must be between 0 and 100");
        }

        // Validate milestones
        if (request.getMilestones() == null || request.getMilestones().isEmpty()) {
            throw new RuntimeException("At least one milestone is required");
        }

        // Validate that end date is after start date for all milestones
        for (ProjectPhaseRequest.MilestoneRequest milestone : request.getMilestones()) {
            // Validate milestone dates are within project timeline
            LocalDate projectStart = project.getCreatedDate().toLocalDate();
            LocalDate projectEnd = project.getEndDate();

            if (milestone.getStartDate().isBefore(projectStart)) {
                throw new RuntimeException("Milestone start date cannot be before project start date");
            }
            if (milestone.getEndDate().isAfter(projectEnd)) {
                throw new RuntimeException("Milestone end date cannot be after project end date");
            }

            if (milestone.getEndDate().isBefore(milestone.getStartDate())) {
                throw new RuntimeException("Milestone end date must be after start date");
            }

            // Validate activities
            if (milestone.getActivities() == null || milestone.getActivities().isEmpty()) {
                throw new RuntimeException("Each milestone must have at least one activity");
            }

            // Validate activity dates and weights
            int activityWeightSum = 0;
            for (ProjectPhaseRequest.MilestoneRequest.ActivityRequest activity : milestone.getActivities()) {
                // Validate activity dates are within milestone dates
                if (activity.getStartDate().isBefore(milestone.getStartDate())) {
                    throw new RuntimeException("Activity start date cannot be before milestone start date");
                }
                if (activity.getEndDate().isAfter(milestone.getEndDate())) {
                    throw new RuntimeException("Activity end date cannot be after milestone end date");
                }
                if (activity.getEndDate().isBefore(activity.getStartDate())) {
                    throw new RuntimeException("Activity end date must be after start date");
                }

                activityWeightSum += activity.getActivityWeight();
            }

            if (activityWeightSum > 100) {
                throw new RuntimeException("Sum of activity weights cannot exceed 100");
            }
        }

        // Validate milestone weights sum
        int milestoneWeightSum = request.getMilestones().stream()
                .mapToInt(ProjectPhaseRequest.MilestoneRequest::getMilestoneWeight)
                .sum();
        if (milestoneWeightSum > 100) {
            throw new RuntimeException("Sum of milestone weights cannot exceed 100");
        }

        // Create phase
        ProjectPhase phase = ProjectPhase.builder()
                .project(project)
                .phaseName(request.getPhaseName())
                .phaseWeight(request.getPhaseWeight())
                .status("ACTIVE")
                .build();

        ProjectPhase savedPhase = phaseRepository.save(phase);
        log.info("Phase created for project {}: {}", projectId, request.getPhaseName());

        // Create milestones and activities
        List<PhaseMilestone> milestones = request.getMilestones().stream()
                .map(milestoneReq -> {
                    PhaseMilestone milestone = PhaseMilestone.builder()
                            .phase(savedPhase)
                            .milestoneName(milestoneReq.getMilestoneName())
                            .startDate(milestoneReq.getStartDate())
                            .endDate(milestoneReq.getEndDate())
                            .milestoneWeight(milestoneReq.getMilestoneWeight())
                            .status("ACTIVE")
                            .build();

                    PhaseMilestone savedMilestone = milestoneRepository.save(milestone);

                    // Create activities
                    List<MilestoneActivity> activities = milestoneReq.getActivities().stream()
                            .map(activityReq -> MilestoneActivity.builder()
                                    .milestone(savedMilestone)
                                    .activityName(activityReq.getActivityName())
                                    .activityWeight(activityReq.getActivityWeight())
                                    .startDate(activityReq.getStartDate().atStartOfDay())
                                    .endDate(activityReq.getEndDate().atStartOfDay())
                                    .description(activityReq.getDescription())
                                    .status("ACTIVE")
                                    .build())
                            .map(activityRepository::save)
                            .collect(Collectors.toList());

                    savedMilestone.setActivities(activities);
                    return savedMilestone;
                })
                .collect(Collectors.toList());

        savedPhase.setMilestones(milestones);
        return convertToResponse(savedPhase);
    }

    @Transactional(readOnly = true)
    public List<ProjectPhaseResponse> getPhasesByProject(Long projectId) {
        return phaseRepository.findByProjectId(projectId).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public ProjectPhaseResponse getPhaseById(Long projectId, Long phaseId) {
        ProjectPhase phase = phaseRepository.findByIdAndProjectId(phaseId, projectId)
                .orElseThrow(() -> new RuntimeException("Phase not found"));
        return convertToResponse(phase);
    }

    @Transactional
    public ProjectPhaseResponse updatePhase(Long projectId, Long phaseId, ProjectPhaseRequest request) {
        ProjectPhase phase = phaseRepository.findByIdAndProjectId(phaseId, projectId)
                .orElseThrow(() -> new RuntimeException("Phase not found"));

        phase.setPhaseName(request.getPhaseName());
        phase.setPhaseWeight(request.getPhaseWeight());

        ProjectPhase updatedPhase = phaseRepository.save(phase);
        log.info("Phase updated: {}", phaseId);

        return convertToResponse(updatedPhase);
    }

    @Transactional
    public void deletePhase(Long projectId, Long phaseId) {
        ProjectPhase phase = phaseRepository.findByIdAndProjectId(phaseId, projectId)
                .orElseThrow(() -> new RuntimeException("Phase not found"));

        phaseRepository.delete(phase);
        log.info("Phase deleted: {}", phaseId);
    }

    private ProjectPhaseResponse convertToResponse(ProjectPhase phase) {
        return ProjectPhaseResponse.builder()
                .id(phase.getId())
                .projectId(phase.getProject().getId())
                .phaseName(phase.getPhaseName())
                .phaseWeight(phase.getPhaseWeight())
                .status(phase.getStatus())
                .createdDate(phase.getCreatedDate())
                .updatedDate(phase.getUpdatedDate())
                .milestones(phase.getMilestones() != null ? phase.getMilestones().stream()
                        .map(this::convertMilestoneToResponse)
                        .collect(Collectors.toList()) : null)
                .build();
    }

    private ProjectPhaseResponse.MilestoneResponse convertMilestoneToResponse(PhaseMilestone milestone) {
        return ProjectPhaseResponse.MilestoneResponse.builder()
                .id(milestone.getId())
                .phaseId(milestone.getPhase().getId())
                .milestoneName(milestone.getMilestoneName())
                .startDate(milestone.getStartDate())
                .endDate(milestone.getEndDate())
                .milestoneWeight(milestone.getMilestoneWeight())
                .status(milestone.getStatus())
                .createdDate(milestone.getCreatedDate())
                .updatedDate(milestone.getUpdatedDate())
                .activities(milestone.getActivities() != null ? milestone.getActivities().stream()
                        .map(this::convertActivityToResponse)
                        .collect(Collectors.toList()) : null)
                .build();
    }

    private ProjectPhaseResponse.MilestoneResponse.ActivityResponse convertActivityToResponse(MilestoneActivity activity) {
        return ProjectPhaseResponse.MilestoneResponse.ActivityResponse.builder()
                .id(activity.getId())
                .milestoneId(activity.getMilestone().getId())
                .activityName(activity.getActivityName())
                .activityWeight(activity.getActivityWeight())
                .status(activity.getStatus())
                .description(activity.getDescription())
                .createdDate(activity.getCreatedDate())
                .updatedDate(activity.getUpdatedDate())
                .build();
    }
}

File: src/services/projectPhaseService.ts
import { authService } from './authService';

const API_BASE_URL = "http://localhost:7080/api";

export interface ProjectPhaseRequest {
  phaseName: string;
  phaseWeight: number;
  milestones: MilestoneRequest[];
}

export interface MilestoneRequest {
  id?: number;
  milestoneName: string;
  startDate: string; // ISO date format
  endDate: string;   // ISO date format
  revisedEndDate?: string;
  milestoneWeight: number;
  milestoneOrder?: number;
  activities: ActivityRequest[];
}

export interface ActivityRequest {
  id?: number;
  activityName: string;
  activityWeight: number;
  description?: string;
  startDate: string; // ISO date format
  endDate: string;   // ISO date format
  revisedEndDate?: string;
}

export interface ProjectPhaseResponse {
  id: number;
  phaseName: string;
  phaseWeight: number;
  status: string;
  createdDate: string;
  updatedDate?: string;
  milestones?: MilestoneResponse[];
}

export interface MilestoneResponse {
  id: number;
  phaseId: number;
  milestoneName: string;
  startDate: string;
  endDate: string;
  revisedEndDate?: string;
  milestoneWeight: number;
  milestoneOrder: number;
  status: string;
  createdDate: string;
  updatedDate?: string;
  activities?: ActivityResponse[];
}

export interface ActivityResponse {
  id: number;
  milestoneId: number;
  activityName: string;
  activityWeight: number;
  status: string;
  description?: string;
  startDate: string;
  endDate: string;
  revisedEndDate?: string;
  createdDate: string;
  updatedDate?: string;
}

class ProjectPhaseService {
  private getHeaders() {
    // Get token from authService first, fallback to localStorage
    let token = authService.getToken();
    
    // If no token from authService, try to get it from localStorage directly
    if (!token) {
      const user = localStorage.getItem('user');
      if (user) {
        try {
          const userData = JSON.parse(user);
          token = userData.token;
        } catch (e) {
          console.error('Error parsing user from localStorage:', e);
        }
      }
    }

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    
    if (token) {
      headers["Authorization"] = `Bearer ${token}`;
    } else {
      console.warn('No JWT token found for API request');
    }
    
    return headers;
  }

  async createPhase(request: ProjectPhaseRequest): Promise<ProjectPhaseResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/project-phases-generic`, {
        method: "POST",
        headers: this.getHeaders(),
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to create phase`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error creating phase:", error);
      throw error;
    }
  }

  async getPhasesByProject(): Promise<ProjectPhaseResponse[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/project-phases-generic`, {
        method: "GET",
        headers: this.getHeaders(),
      });

      if (!response.ok) {
        throw new Error("Failed to fetch phases");
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching phases:", error);
      throw error;
    }
  }

  async getPhaseById(phaseId: number): Promise<ProjectPhaseResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/project-phases-generic/${phaseId}`, {
        method: "GET",
        headers: this.getHeaders(),
      });

      if (!response.ok) {
        throw new Error("Failed to fetch phase");
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching phase:", error);
      throw error;
    }
  }

  async updatePhase(phaseId: number, request: ProjectPhaseRequest): Promise<ProjectPhaseResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/project-phases-generic/${phaseId}`, {
        method: "PUT",
        headers: this.getHeaders(),
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to update phase`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error updating phase:", error);
      throw error;
    }
  }

  async deletePhase(phaseId: number): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/project-phases-generic/${phaseId}`, {
        method: "DELETE",
        headers: this.getHeaders(),
      });

      if (!response.ok) {
        throw new Error("Failed to delete phase");
      }
    } catch (error) {
      console.error("Error deleting phase:", error);
      throw error;
    }
  }
}

export const projectPhaseService = new ProjectPhaseService();
